### ISSUE TITLE
Unable to install into system: writing LaunchAgent plist: open ... permission denied

### ISSUE THREAD

**Issue by haggen** [puma/puma-dev#141]
Clean install of High Sierra, I ran `sudo puma-dev -setup` which worked, then `puma-dev -install -d local` which output:

```
* Use '/usr/local/Cellar/puma-dev/0.11/bin/puma-dev' as the location of puma-dev
2017/12/13 15:15:24 Unable to install into system: writing LaunchAgent plist: open /Users/Arthur/Library/LaunchAgents/io.puma.dev.plist: permission denied
```

**Comment by cforcey** [comment]
I set up a new laptop on 12/28 and went through the High Sierra updates to get to macOS 10.13.2 (17C88).  At that point, I tried to install puma-dev and got the same message above on puma-dev -install -d test (my variation).  

When I ran Get Info on my user folder, I saw an unusual message instead of the usual Sharing and Permission table of users and privileges: "you have unknown access."  I followed the instructions in this apple support article despite the inability to set permissions on folders as instructed and ran the repair tasks.  No luck!

I rebooted in case I was in some liminal state after all the updates -- there was some weird App Store showing and then hiding a recent security update after telling me it no longer applied -- and after the reboot I got the expect table of permissions on Get Info and was able to follow these instructions to the letter.  When completed, puma-dev -install -d test worked as expected.

https://support.apple.com/en-us/HT203538

Hope this helps -- additional insights and explanations welcome.

**Comment by haggen** [comment]
I'll look into that, thank you @cforcey.

**Comment by evanphx** [comment]
Yeah, that's configuration your machine, perhaps due to security clamping done in later OS X versions?

**Comment by dbackeus** [comment]
This error is still occurring. Just got it with a fresh reinstall of MacOS 10.15.4. Reboot didn't help.

**Comment by nonrational** [comment]
@dbackeus can you provide some debugging information?

1. What version of puma-dev are you using? 
1. Check existence / permissions of that file: `ls -la $HOME/Library/LaunchAgents/io.puma.dev.plist`
1. Check existence / permissions of the parent directory: `ls -la $HOME/Library/LaunchAgents/`

**Comment by dbackeus** [comment]
Latest version of puma-dev.

Somehow the issue appears to be weird permissions on the `LaunchAgents` directory. Even though `ls -la` showed `drw-r--r--    2 davidbackeus  staff    64 Apr 17 20:57 LaunchAgents` indicating read + write for my user when I opened the Finder info window for that folder and the permissions were listed as "Custom" for my user / group and everyone. I manually clicked the dropdown for my own user and changed it to "Read & Write" which fixed the issue. Unfortunately I didn't take a screenshot of this for fixing and I don't know how to revert back to the "Custom" state.

Now `ls -la` shows `drwxr--r--    5 davidbackeus  staff    160 Apr 17 22:22 LaunchAgents` indicating `x` was added to my user. But I don't see how that should be necessary for writing a file?

**Comment by nonrational** [comment]
Over-simplified answer: You need `+x` to be able to enter a directory and interact with its contents.

https://github.com/puma/puma-dev/blob/2163f317517d51dec5d22a3fffce880541e54d70/dev/setup.go#L162

A cursory look at the go code, looks like we create that directory if it doesn't exist and ... drumroll ... do so with the _wrong_ permissions!

**Comment by dbackeus** [comment]
That makes sense then. Can confirm that I ran this as the first thing happening on a brand new MacOS install so likely the directory didn't exist. I can at least confirm that the folder was empty at the time of getting the error.

**Comment by MikeRogers0** [comment]
I setup a fresh install today, I saw the following:

```
$ puma-dev -install -d puma                                                                                                                                                                                                                    
* Adding certification to login keychain as trusted
! There is probably a dialog open that requires you to authenticate
* Certificates setup, ready for https operations!
* Use '/usr/local/Cellar/puma-dev/0.13/bin/puma-dev' as the location of puma-dev
2020/04/18 21:56:19 Unable to install into system: writing LaunchAgent plist: open /Users/mike/Library/LaunchAgents/io.puma.dev.plist: permission denied
```

I was able to solve it by running:

```
cd ~/Libary && chmod 0755 LaunchAgents  
```

**Comment by nonrational** [comment]
yup! that'll do it. 

i've got a branch that fixes the issue going forward, but currently trying to figure out how to test it without blowing away `~/LaunchAgents/` every time.

**Comment by dbackeus** [comment]
Would appreciate if this could get properly released as well. Just got the same error on a new computer today.

**Comment by nonrational** [comment]
I will cut a v0.0.14 for homebrew next week. Was hoping to get some keychain stuff landed, but it's proving tricky. It'll have to wait for v0.0.15. 

Thanks for the polite bump.

**Comment by nonrational** [comment]
This is now released as part of v0.15. Homebrew formula updated.
